### DISCUSSION TITLE
Python API

### DISCUSSION THREAD

**Discussion by Jade-1s** [eclipse/mosaic#330] [General]
Hi all, 

I am wondering whether there is a plan to develop Python API for implementing applications. I noticed that we need to write Java for applications, but as SUMO users, I usually write Python with the help of Traci. If Mosaic also provides API like Traci, it will help a lot for integrating Mosaic into SUMO.

**Comment by schwepmo** [comment]
Currently there are no plans to develop a Python API. MOSAIC uses a custom TraCI coupling using the socket interface or Libsumo to communicate with SUMO. 
I also think you are misunderstanding the roles of Eclipse SUMO and MOSAIC a little bit.
MOSAIC is a co-simulation framework that couples many simulators from different domains, while SUMO is a traffic simulator that can be co-simulated by MOSAIC allowing for example to simulate V2X communication between vehicles simulated by SUMO.

**Comment by nemcikjan** [comment]
@schwepmo maybe a better question would be if you are planning to add a possibility to write Mosaic applications in Python? It'd mean, that Application Simulator component would need to be able to simulate Python applications. Since Python is completely different language than Java, is this even possible?

**Reply by schwepmo** [reply]
There are currently no plans/resources to realize this internally. We had an initial test using a grpc-based abstraction layer between applications written in different languages (e.g., Python) and the Application Simulator, which allowed accessing basic functionalities at a little bit of performance loss. However, we didn't pursue this any further.

Feel free to open an enhancement issue and if we notice a high demand for any form of Python API, we may try to get something working in the future.

**Reply by nemcikjan** [reply]
To be honest, I'd prefer grpc over python specific API, since the integration possibilities will be much wider. Since you had tested grpc pilot, is there any existing issue for this or should I create a new enhancement issue as you suggested.

**Reply by schwepmo** [reply]
You can go ahead and create a new issue. The prototype we build wasn't tracked in github at all.
